cluster-autoscaler/core/scaledown/unneeded/nodes.go

```diff
@@ -0,0 +1,213 @@
+
+	apiv1 "k8s.io/api/core/v1"
+	klog "k8s.io/klog/v2"
+)
+
+// Nodes tracks the state of cluster nodes that are not needed.
+type Nodes struct {
+	sdtg         scaleDownTimeGetter
+	limitsFinder *resource.LimitsFinder
+	asList       []*apiv1.Node
+	byName       map[string]*node
+}
+
+type node struct {
+	ntbr  simulator.NodeToBeRemoved
+	since time.Time
+}
+
+type scaleDownTimeGetter interface {
+	// GetScaleDownUnneededTime returns ScaleDownUnneededTime value that should be used for a given NodeGroup.
+	GetScaleDownUnneededTime(context *context.AutoscalingContext, nodeGroup cloudprovider.NodeGroup) (time.Duration, error)
+	// GetScaleDownUnreadyTime returns ScaleDownUnreadyTime value that should be used for a given NodeGroup.
+	GetScaleDownUnreadyTime(context *context.AutoscalingContext, nodeGroup cloudprovider.NodeGroup) (time.Duration, error)
+}
+
+// NewNodes returns a new initialized Nodes object.
+func NewNodes(sdtg scaleDownTimeGetter, limitsFinder *resource.LimitsFinder) *Nodes {
+	return &Nodes{
+		sdtg:         sdtg,
+		limitsFinder: limitsFinder,
+	}
+}
+
+// Update add stores nodes along with a time at which they were found to be
+// unneeded. Previously existing timestamps are preserved.
+func (n *Nodes) Update(nodes []simulator.NodeToBeRemoved, ts time.Time) {
+	updated := make(map[string]*node, len(nodes))
+	for _, nn := range nodes {
+		name := nn.Node.Name
+		updated[name] = &node{
+			ntbr: nn,
+		}
+		if val, found := n.byName[name]; found {
+			updated[name].since = val.since
+		} else {
+			updated[name].since = ts
+		}
+	}
+	n.byName = updated
+	n.asList = make([]*apiv1.Node, 0, len(nodes))
+	for _, v := range n.byName {
+		n.asList = append(n.asList, v.ntbr.Node)
+	}
+	if klog.V(4).Enabled() {
+		for k, v := range n.byName {
+			klog.Infof("%s is unneeded since %s duration %s", k, v.since, ts.Sub(v.since))
+		}
+	}
+}
+
+// Clear resets the internal state, dropping information about all tracked nodes.
+func (n *Nodes) Clear() {
+	n.Update(nil, time.Time{})
+}
+
+// Contains returns true iff a given node is unneeded.
+func (n *Nodes) Contains(nodeName string) bool {
+	_, found := n.byName[nodeName]
+	return found
+}
+
+// AsList returns a slice of unneeded Node objects.
+func (n *Nodes) AsList() []*apiv1.Node {
+	return n.asList
+}
+
+// Drop stops tracking a specified node.
+// List representation (as returned by AsList) will keep including this node until the next Update().
+func (n *Nodes) Drop(node string) {
+	delete(n.byName, node)
+}
+
+// RemovableAt returns all nodes that can be removed at a given time, divided
+// into empty and non-empty node lists, as well as a list of nodes that were
+// unneeded, but are not removable, annotated by reason.
+func (n *Nodes) RemovableAt(context *context.AutoscalingContext, ts time.Time, resourcesLeft resource.Limits, resourcesWithLimits []string, as scaledown.ActuationStatus) (empty, needDrain []*apiv1.Node, unremovable []*simulator.UnremovableNode) {
```

Comment:
I'm not really sure if this function should be responsible for dividing nodes into empty/non-empty:
 * The checks performed by Nodes() are all related to node itself and not pods running on it (which after your refactor is the domain of eligibility I think?). The fact that this package is grouping nodes by pods here seems inconsistent with what the rest of this package is doing and, even more so, outside of what I'd expect to be the responsibility of this function.
 * The grouping is done based on cached result passed to Update(). In current implementation that's fresh result, but that's a pretty big assumption. Obviously we want to avoid stale data leading to node being incorrectly considered empty.
 * Current implementation doesn't use that grouping anyway (it puts both groups into one list) and relies on re-checking this in getEmptyNodesToRemove(). Do we expect new scale-down implementation to act differently?

If we drop this groupby we should also update this to cache nodes and not ntbr, since as far as I can tell that's the only place where we use ntbr.

Reply:
I actually planned to avoid running the simulation twice in new scale down implementation, so the split done here would be final. We can get it wrong, and sometimes already do, which actually proves re-checking doesn't solve the problem, just reduces the chance of a race condition. The proper solution, I think, is to first apply the taints and only then look at the state of the cluster to determine which ones are empty. This was discussed and should be done in https://github.com/kubernetes/autoscaler/issues/3886. If you disagree we can discuss it there.